snuba/query/expressions.py

```diff
@@ -0,0 +1,215 @@
+        transformation function and we do not run that same function over the
+        new children.
+        """
+        transformed = self._duplicate_with_new_children(
+            list(map(lambda child: child.transform(func), self._get_children()))
+        )
+        return func(transformed)
+
+    def __iter__(self) -> Iterator[Expression]:
+        """
+        Traverse the subtree in a postfix order.
+        The order here is arbitrary, postfix is chosen to follow the same
+        order we have in the transform method.
+        """
+        for child in self._get_children():
+            for sub in child:
+                yield sub
+        yield self
+
+    @abstractmethod
+    def _get_children(self) -> Sequence[Expression]:
+        """
+        To be implemented by the subclasses to provide the list of children
+        for iteration/transformation.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _duplicate_with_new_children(self, children: Sequence[Expression]) -> Expression:
+        """
+        Return a new instance of the expression with a new set of children.
+        """
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class Literal(Expression):
+    """
+    A literal in the SQL expression
+    """
+    value: str
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        return func(self)
+
+    def __iter__(self) -> Iterator[Expression]:
+        yield self
+
+
+@dataclass(frozen=True)
+class Column(Expression):
+    """
+    Represent a column in the schema of the dataset.
+    """
+    column_name: str
+    table_name: Optional[str]
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        return func(self)
+
+    def __iter__(self) -> Iterator[Expression]:
+        yield self
+
+
+@dataclass(frozen=True)
+class FunctionCall(HierarchicalExpression):
+    """
+    Represents an expression that resolves to a function call on Clickhouse.
+    This class also represent conditions. Since Clickhouse supports both the conventional
+    infix notation for condition and the functional one, we converge into one
+    representation only in the AST to make query processing easier.
+    A query processor would not have to care of processing both functional conditions
+    and infix conditions.
+    """
+    function_name: str
+    parameters: Sequence[Expression]
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def _get_children(self) -> Sequence[Expression]:
+        return self.parameters
+
+    def _duplicate_with_new_children(self, children: Sequence[Expression]) -> Expression:
+        return FunctionCall(self.alias, self.function_name, children)
+
+
+@dataclass(frozen=True)
+class Aggregation(HierarchicalExpression):
```

Comment:
Would it make sense to make `Aggregation` a subtype of `FunctionCall`? Or do you think it's better to keep them siblings in the type hierarchy? Seems to me that might be the best of both worlds for both consistency as well as the ability to do `isinstance(expr, Aggregation)` for now.

This would also make `HierarchicalExpression` unnecessary (of which I still think removing would make the design easier to understand.)